examples/defaulter-gen/generators/defaulter.go

```diff
@@ -424,32 +447,43 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
 		// if the base function indicates it "covers" (it already includes defaulters)
 		// we can halt recursion
 		if checkTag(defaults.base.CommentLines, "covers") {
-			glog.V(6).Infof("the defaulter %s indicates it covers all sub generators", t.Name)
+			glog.V(6).Infof("the defaulter %s indicates it covers all sub generators", c.t.Name)
 			return parent
 		}
 	}
 
 	// base has been added already, now add any additional defaulters defined for this object
 	parent.call = append(parent.call, defaults.additional...)
 
-	switch t.Kind {
+	// if the type already exists, don't build the tree for it and don't generate anything.
+	// This is used to avoid recursion for nested recursive types.
+	if c.currentlyBuildingTypes[c.t] {
+		return nil
+	}
+	// if type doesn't exist, mark it as existing
+	c.currentlyBuildingTypes[c.t] = true
+
+	switch c.t.Kind {
 	case types.Pointer:
-		if child := buildCallTreeForType(t.Elem, false, existingDefaulters, newDefaulters); child != nil {
+		r := NewCallTreeForType(c.t.Elem, false, c.existingDefaulters, c.newDefaulters, c.currentlyBuildingTypes)
+		if child := r.buildCallTreeForType(); child != nil {
```

Comment:
All of these should look like `if child := r.Build(c.t.Elem, false); child != nil {`